### PR TITLE
Idempotency

### LINKED ISSUE #160: Remove v1 endpoint
webhookHandler has been migrated to v2, time to remove v1 once we are sure everything is stable.

### LINKED ISSUE #161: Idempotency
Currently the userHandler is not idempotent, and if it receives duplicates (whatsapp message id is the same) it will create multiple instances and/or messages.

Need to change to make it idempotent.

### PR DESCRIPTION
**Pull request by sarge1989**
Resolves #161
Resolves #160

## Summary
- Created a new db collection just to store incoming user bot wamids.
- Now messages and instances are created in a transaction, together with the document in the abovementioned wamid collection. Before that, the transasction gets the document with that wamid from the wamid collection, preventing duplicates messages/instances from being created. [Firebase transaction documentation](https://firebase.google.com/docs/firestore/manage-data/transactions#transactions).
- Also checks for presence of same wamid at webhook handler

## Notes
- Am not doing the same for checkers, since migrating out whatsapp soon anyway. Impact of non-idempotency is also less.

### CODE REVIEW

File: functions/src/definitions/userHandlers.ts [hunk 1]
```diff
@@ -792,6 +828,33 @@ async function onTextListReceipt(messageObj: Message, platform = "whatsapp") {
...
+  instanceUpdateObj: Object
+) {
+  const messageIdRef = db.collection("messageIds").doc(id)
+  try {
+    await db.runTransaction(async (t) => {
+      const doc = await t.get(messageIdRef)
+      if (doc.exists) {
+        return
+      }
+      if (!hasMatch && !!messageRef && !!messageUpdateObj) {
```
Comment (stonefruit) [inline functions/src/definitions/userHandlers.ts:846]: Just a note. For `if` conditions, you don't have to use `!!` actually. If assigning boolean values, then the `!!` is necessary.

File: functions/src/definitions/common/utils.ts [hunk 2]
```diff
@@ -33,6 +33,13 @@ const checkUrl = function (urlString: string) {
   return url.protocol === "http:" || url.protocol === "https:"
 }
 
+async function checkMessageId(messageId: string) {
+  const db = admin.firestore()
+  const messageRef = db.doc(`messagesIds/${messageId}`)
```
Comment (stonefruit) [inline functions/src/definitions/common/utils.ts:38]: Not sure if this is the correct syntax as the samples in firebase docs seems to require creating a similar to `const messageIdRef = db.collection("messageIds").doc(id)` found below.

Also, `message(s)Ids` vs `messageId` looks suspicious

File: functions/src/definitions/userHandlers.ts [hunk 3]
```diff
@@ -792,6 +828,33 @@ async function onTextListReceipt(messageObj: Message, platform = "whatsapp") {
...
 
+async function addInstanceToDb(
+  id: string,
+  hasMatch: boolean,
+  messageRef: FirebaseFirestore.DocumentReference | null,
+  messageUpdateObj: Object | null = null,
+  instanceRef: FirebaseFirestore.DocumentReference,
+  instanceUpdateObj: Object
+) {
+  const messageIdRef = db.collection("messageIds").doc(id)
```
Comment (stonefruit) [inline functions/src/definitions/userHandlers.ts:839]: Related to first comment

**Review by stonefruit — CHANGES_REQUESTED** [review]
Just want to check on the messageId reference comment

File: functions/src/definitions/userHandlers.ts (hunk 1 above)
Comment (sarge1989) [reply functions/src/definitions/userHandlers.ts:846]: Noted!

File: functions/src/definitions/common/utils.ts (hunk 2 above)
Comment (sarge1989) [reply functions/src/definitions/common/utils.ts]: Good catch, fixed!

**Review by stonefruit — APPROVED** [review]

**Pull request merged**